prov/util/src/util_wait.c

```diff
@@ -424,3 +428,210 @@ int ofi_wait_fd_open(struct fid_fabric *fabric_fid, struct fi_wait_attr *attr,
 	free(wait);
 	return ret;
 }
+
+static void util_wait_yield_signal(struct util_wait *util_wait)
+{
+	struct util_wait_yield *wait_yield;
+
+	wait_yield = container_of(util_wait, struct util_wait_yield, util_wait);
+
+	fastlock_acquire(&wait_yield->signal_lock);
+	wait_yield->signal = 1;
+	fastlock_release(&wait_yield->signal_lock);
+}
+
+static int util_wait_yield_run(struct fid_wait *wait_fid, int timeout)
+{
+	struct util_wait_yield *wait = container_of(wait_fid,
+			struct util_wait_yield, util_wait.wait_fid);
+	struct ofi_wait_fid_entry *fid_entry;
+	int ret = 0;
+
+	fastlock_acquire(&wait->wait_lock);
+	while (!wait->signal) {
+		dlist_foreach_container(&wait->fid_list,
+					struct ofi_wait_fid_entry,
+					fid_entry, entry) {
+			ret = fid_entry->wait_try(fid_entry->fid);
+			if (ret)
+				return ret;
+		}
+	}
+	fastlock_release(&wait->wait_lock);
+
+	fastlock_acquire(&wait->signal_lock);
+	wait->signal = 0;
+	fastlock_release(&wait->signal_lock);
+
+	return FI_SUCCESS;
+}
+
+static int util_wait_yield_close(struct fid *fid)
+{
+	struct util_wait_yield *wait;
+	struct ofi_wait_fid_entry *fid_entry;
+	int ret;
+
+	wait = container_of(fid, struct util_wait_yield, util_wait.wait_fid.fid);
+	ret = fi_wait_cleanup(&wait->util_wait);
+	if (ret)
+		return ret;
+
+	fastlock_acquire(&wait->wait_lock);
+	while (!dlist_empty(&wait->fid_list)) {
+		dlist_pop_front(&wait->fid_list, struct ofi_wait_fid_entry,
+				fid_entry, entry);
+		free(fid_entry);
+	}
+	fastlock_release(&wait->wait_lock);
```

Comment:
The locking here isn't needed.  We're about to free the lock and wait itself, so this lock can't provide any protection.